### PR TITLE
refactor(redirect): make into self-contained module

### PR DESCRIPTION
**Pull request by LoneRifle**
## Context

As part of a wider effort to reorganise the codebase around
domain-specific modules, consolidate all redirect-related
objects into a redirect module, under the `redirect/` dir.

TODO: Move tests to under src, move link statistics logging to
AnalyticsLoggerService

Ongoing work for #869 

## Solution

- create a new redirect dir, for the redirect module
- ~create a Redirect class in `redirect/index.ts`, holding the
  redirect endpoint and the endpoint for gtag.js, and accessible
  via `Redirect.router`. Make this injectable via inversify~ discarded in favour of injecting the controller directly

- ~Rework `server/index.ts` to mount `Redirect.router` on top of
  `redirectSpecificMiddleware`, so that gtag.js endpoint now
  has the session cookie. This is deliberate to better organise
  the codebase~

- Move routing, controller, services, types to `redirect/`
- GaLogger -> AnalyticsLogger
- Remove all interfaces for redirection

### Improvements

Under this new module-centric scheme, the `index.ts` of the module 
will export by default something that is mountable into express, either 
a router or a controller. The index file will also export relevant controllers
and types by name.

Also take the trouble to dispense with interfaces. By and large, we
no longer have to use interfaces to specify a contract in TypeScript
[[1]]. Moreover, because we don't actually have multiple implementations
of the interfaces we have declared, we can just use classes, and then
mock the class using the `Required<T>` utility type in TypeScript [[2]]

[1]: https://docs.microsoft.com/en-gb/archive/blogs/kcwalina/api-design-myth-interface-as-contract
[2]: https://stackoverflow.com/questions/31876947/confused-about-the-interface-and-class-coding-guidelines-for-typescript

### CODE REVIEW

**Comment by LoneRifle** [comment]
Failure is due to a decrease in code coverage owing to a reorg

File: src/server/inversify.config.ts [hunk 1]
```diff
@@ -161,4 +143,6 @@ export default () => {
   }
 
   bindIfUnbound(DependencyIds.s3, S3ServerSide)
+
+  addRedirectBindings(container)
```
Comment (liangyuanruo) [inline src/server/inversify.config.ts:147]: i'm still trying to understand the value of this abstraction, given that all the other bindings are performed inline above, and other developers can tell at a glance what's happening. is it supposed to be the start of introducing a loader pattern? but then, the implementation is specific to the redirect module.

it's only partially abstracting away DI, which is an edge case. can we leave it out as a separate PR until we're ready to tackle the entire thing?

**Review by liangyuanruo — CHANGES_REQUESTED** [review]
i'd like to prioritise ease of reading of the codebase. i think `addRedirectBindings` is introducing complexity rather than removing it. if you want to introduce the `makeRouter` pattern, do it at a later stage in a separate PR?

**Comment by LoneRifle** [comment]
> i'd like to prioritise ease of reading of the codebase. i think `addRedirectBindings` is introducing complexity rather than removing it. if you want to introduce the `makeRouter` pattern, do it at a later stage in a separate PR?

Is this a change in your position? Previous discussions indicate you were alright with having the module handle its own inversify dep injection

File: test/server/modules/redirect/RedirectController.test.ts [hunk 2]
```diff
@@ -152,31 +162,52 @@ describe('redirect API tests', () => {
...
+  test('url exists in cache and db, real user visited without cookie', async () => {
+    container.unbind(DependencyIds.cookieReducer)
+    container
+      .bind<CookieArrayReducerService>(DependencyIds.cookieReducer)
+      .toConstantValue(cookieArrayReducerMockVisited)
+    redisMockClient.set('aaa', 'aa')
+
+    const req = createRequestWithShortUrl('Aaa')
+    const res = httpMocks.createResponse()
+    req.headers['user-agent'] = userAgent
```
[the remark this answers is not in the document]
Comment (LoneRifle) [reply test/server/modules/redirect/RedirectController.test.ts:210]: Out of scope of the commit? I would have thought the priority would be reorganisation

**Review by liangyuanruo — APPROVED** [review]
> Is this a change in your position? Previous discussions indicate you were alright with having the module handle its own inversify dep injection

As discussed, let's look into performing this refactor in a separate issue.

**Pull request merged**